alf/utils/image_networks.py

```diff
@@ -0,0 +1,216 @@
+
+    def __init__(self,
+                 conv_layer_params,
+                 padding="valid",
+                 postprocess_fc_layer_params=None,
+                 activation_fn=tf.nn.relu,
+                 output_activation_fn=None,
+                 name="image_encoding_network"):
+        """
+        Initialize the layers for encoding an image into a latent vector.
+
+        Args:
+            conv_layer_params (list[tuple]): a non-empty list of elements
+                (num_filters, kernel_size, strides).
+            padding (str): "same" or "valid", see tf.keras.layers.Conv2DTranspose
+            postprocess_fc_layer_params (tuple[int]): a list of fc layer units.
+                These fc layers are used for postprocessing the conv output. If
+                None, then the output would preserve the shape of an image (will
+                not be flattened).
+            activation_fn (tf.nn.activation): activation for conv layers
+            output_activation_fn (tf.nn.activation): only used when there are
+                postprocessing FC layers
+            name (str): network name
+        """
+        super().__init__(name=name)
+
+        assert isinstance(conv_layer_params, list)
+        assert len(conv_layer_params) > 0
+
+        self._postprocess_fc_layers = []
+        if postprocess_fc_layer_params is not None:
+            for size in postprocess_fc_layer_params[:-1]:
+                self._postprocess_fc_layers.append(
+                    tf.keras.layers.Dense(size, activation=activation_fn))
+            self._postprocess_fc_layers.append(
+                tf.keras.layers.Dense(
+                    postprocess_fc_layer_params[-1],
+                    activation=output_activation_fn))
+
+        self._postprocess_fc_layer_params = postprocess_fc_layer_params
+        self._conv_layer_params = conv_layer_params
+        self._padding = padding
+        self._conv_layers = []
+        for filters, kernel_size, strides in conv_layer_params:
+            self._conv_layers.append(
+                tf.keras.layers.Conv2D(
+                    padding=padding,
+                    filters=filters,
+                    kernel_size=kernel_size,
+                    strides=strides,
+                    activation=activation_fn))
+
+    def output_shape(self, input_size, conv=False):
+        """Return the output shape given the input image size.
+
+        How to calculate the output size:
+
+            if padding=="same", then H = (H1 - 1) // strides + 1
+            if padding=="valid", then H = (H1 - HF) // strides + 1
+        where H = output size, H1 = input size, HF = size of kernel
+
+        Args:
+            input_size (int): the input image size
+            conv (bool): whether calculate the intermediate conv output size
+                or the final FC output size
+
+        Returns:
+            a tuple representing the output shape
+        """
+        if self._postprocess_fc_layers and not conv:
```

Comment:
What is the corresponding scenario when self._postprocess_fc_layer_params is not None and conv is True?

Reply:
It will return the last conv's shape, regardless of postprocess FC layers